File: thor/src/jvmMain/kotlin/io/github/remmerw/thor/cobra/util/Items.java
package io.github.remmerw.thor.cobra.util;

import java.util.HashMap;
import java.util.Map;
import java.util.WeakHashMap;

public class Items {
    private static final Map<Object, Map<String, Object>> sourceMap = new WeakHashMap<>();

    private Items() {
    }

    public static Object getItem(final Object source, final String name) {
        final Map<Object, Map<String, Object>> sm = sourceMap;
        synchronized (sm) {
            final Map<String, Object> itemMap = sm.get(source);
            if (itemMap == null) {
                return null;
            }
            return itemMap.get(name);
        }
    }

    public static void setItem(final Object source, final String name, final Object value) {
        final Map<Object, Map<String, Object>> sm = sourceMap;
        synchronized (sm) {
            Map<String, Object> itemMap = sm.get(source);
            if (itemMap == null) {
                itemMap = new HashMap<>(1);
                sm.put(source, itemMap);
            }
            itemMap.put(name, value);
        }
    }
}

File: thor/src/jvmMain/kotlin/io/github/remmerw/thor/cobra/util/Items.kt
package io.github.remmerw.thor.cobra.util

import java.util.WeakHashMap

object Items {
    private val sourceMap: MutableMap<Any?, MutableMap<String?, Any?>?> =
        WeakHashMap<Any?, MutableMap<String?, Any?>?>()

    fun getItem(source: Any?, name: String?): Any? {
        val sm = sourceMap
        synchronized(sm) {
            val itemMap = sm.get(source)
            if (itemMap == null) {
                return null
            }
            return itemMap.get(name)
        }
    }

    fun setItem(source: Any?, name: String?, value: Any?) {
        val sm = sourceMap
        synchronized(sm) {
            var itemMap = sm.get(source)
            if (itemMap == null) {
                itemMap = HashMap<String?, Any?>(1)
                sm.put(source, itemMap)
            }
            itemMap.put(name, value)
        }
    }
}
